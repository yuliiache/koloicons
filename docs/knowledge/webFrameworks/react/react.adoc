== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







