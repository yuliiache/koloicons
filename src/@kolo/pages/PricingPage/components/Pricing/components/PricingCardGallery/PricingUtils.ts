import {LabelTheme} from '@kolo/uiKit/Label/constants';

import {Currency, SubscriptionPlan} from '../../constants';

const LabelThemes = {
  [SubscriptionPlan.PERSONAL]: LabelTheme.SUCCESS,
  [SubscriptionPlan.BUSINESS]: LabelTheme.PRIMARY,
  [SubscriptionPlan.EXTENDED]: LabelTheme.WARNING,
};

export const getThemeForPlan = (plan: SubscriptionPlan): LabelTheme => {
  return LabelThemes[plan] || LabelTheme.DEFAULT;
};

const CurrencySymbol = {
  [Currency.USD]: '$',
  [Currency.EUR]: 'â‚¬',
} as const;

export const formatPrice = (price: number, currencyCode: Currency): string => {
  const currencySymbol = CurrencySymbol[currencyCode];
  const formattedPrice = price.toFixed(2);

  return `${currencySymbol}${formattedPrice}`;
};
