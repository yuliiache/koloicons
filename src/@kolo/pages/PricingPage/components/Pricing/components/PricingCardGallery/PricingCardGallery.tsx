import PricingCard from '@kolo/uiKit/PricingCard/PricingCard';
import React from 'react';

import styles from '../../Pricing.module.scss';
import {PricingData} from '../../types/pricingStateTypes';
import {formatPrice, getThemeForPlan} from './PricingUtils';

const {pricingCards} = styles;

const PricingCardGallery = ({pricingPlans}: {pricingPlans: PricingData[]}) => {
  return (
    <div className={pricingCards}>
      {pricingPlans.map(({features, license, options, paymentId, subscriptionPlan}) => (
        <PricingCard
          key={paymentId}
          name={subscriptionPlan}
          price={formatPrice(options.price, options.currency)}
          theme={getThemeForPlan(subscriptionPlan)}
          periodInfo={options.period}
          planFeatures={features}
          allowedOptions={license.allowed}
          disallowedOptions={license.disallowed}
        />
      ))}
    </div>
  );
};

export default PricingCardGallery;
