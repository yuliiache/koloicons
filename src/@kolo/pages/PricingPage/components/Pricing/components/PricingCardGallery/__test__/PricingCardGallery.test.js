import React from 'react';
import PricingCardGallery from '@kolo/pages/PricingPage/components/Pricing/components/PricingCardGallery/PricingCardGallery';
import {render} from '@testing-library/react';

import '@testing-library/jest-dom';
const singlePricingPlan = [
  {
    features: ['Feature 1', 'Feature 2'],
    license: {allowed: ['Option 1'], disallowed: ['Option 2']},
    options: {price: 10, currency: 'USD', period: 'month'},
    paymentId: 'plan1',
    subscriptionPlan: 'Basic',
  },
];

describe('PricingCardGallery', () => {
  it('renders a single pricing card for one pricing plan', () => {
    const {asFragment} = render(<PricingCardGallery pricingPlans={singlePricingPlan} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
