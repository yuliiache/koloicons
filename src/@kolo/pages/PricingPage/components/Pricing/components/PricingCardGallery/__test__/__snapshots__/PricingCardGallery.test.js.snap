// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PricingCardGallery renders a single pricing card for one pricing plan 1`] = `
<DocumentFragment>
  <div
    class="pricingCards"
  >
    <div
      class="card"
    >
      <span
        class="label default large"
        data-testid="labelComponent"
      >
        Basic
      </span>
      <div
        class="price"
      >
        <span
          class="priceVal"
        >
          $10.00
        </span>
        <span
          class="priceDesc"
        >
          month
        </span>
      </div>
      <ul
        class="planFeatures"
      >
        <li>
          <strong>
            Feature 1
          </strong>
        </li>
        <li>
          <strong>
            Feature 2
          </strong>
        </li>
      </ul>
      <button
        class="button button-full-width button-primary"
        data-testid="buttonComponent"
        type="button"
      >
        <span
          data-testid="buttonComponentText"
        >
          Purchase
        </span>
      </button>
      <div
        class="advantages"
      >
        <p
          class="advantagesItem"
        >
          <span
            class="itemTitle"
          >
            Formats:
          </span>
          <span
            class="itemVal"
          >
            PNG, SVG
          </span>
        </p>
        <p
          class="advantagesItem"
        >
          <span
            class="itemTitle"
          >
            Attribution:
          </span>
          <span
            class="itemVal"
          >
            No attribution required
          </span>
        </p>
      </div>
      <dl
        class="advantages"
      >
        <dt
          class="optionTitle"
        >
          You can use icons for:
        </dt>
        <dd
          class="optionVal"
        >
          <span>
             - 
          </span>
          Option 1
        </dd>
        <dt
          class="optionTitle"
        >
          You can not use icons for:
        </dt>
        <dd
          class="optionVal"
        >
          <span>
             - 
          </span>
          Option 2
        </dd>
      </dl>
    </div>
  </div>
</DocumentFragment>
`;
